### DISCUSSION TITLE
Is TreeView select/expand state persistent?

### DISCUSSION THREAD

**Discussion by drok** [microsoft/vscode-discussions#1071] [Extension Development QnA] [answered]
It is not clear to me from the TreeView extension guide whether the expansion and selection state of a TreeView is supposed to be persisted by vscode. Is it?

My extension discovers tests and adds them to the Test Explorer. The expansion state of the test tree is persisted by vscode, without my extension doing anything special.

My extension also contributes a TreeView container with two TreeViews. I've paid special attention that every TreeItem has a unique id and label in its respective tree, but the expansion state of my trees is not persisted automatically.

I've also noticed that the File Explorer and the [tree-view-sample](https://github.com/Microsoft/vscode-extension-samples/tree-view-sample) trees also do not have their state persisted.

Is there something I can do to get persistence as implemented in the Test Explorer UI, for the two custom tree views my extension contributes? Must I implement my own persistence algo by saving expansion state from `onDidCollapseElement` and `onDidExpandElement` in the worspace memento state ?

Thank you

**Comment by isidorn** [comment]
@alexr00 might be able to help

**Comment by alexr00** [comment, marked as answer]
Extension tree view expand/collapse state is not persisted. You would need to listen to the `onDidExpandElement` and `onDidCollapseElement` as you describe.
